File: 2/day2.go
package day2

import (
	"os"
	"strconv"
	"strings"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func split(input string, on string) (string, string) {
	_sliced := strings.Split(input, on)
	return _sliced[0], _sliced[1]
}

func getEmptyBag() map[string]int {
	contents := make(map[string]int)

	contents["red"] = 0
	contents["green"] = 0
	contents["blue"] = 0

	return contents
}

// func getBag() map[string]int {
// 	contents := make(map[string]int)

// 	contents["red"] = 12
// 	contents["green"] = 13
// 	contents["blue"] = 14

// 	return contents
// }

// func isValidMove(move string) bool {
// 	bag := getBag()

// 	for _, cubes := range strings.Split(move, ",") {
// 		splitted_cubes := strings.Split(cubes, " ")
// 		_count, color := splitted_cubes[1], splitted_cubes[2]
// 		count, err := strconv.Atoi(_count)
// 		check(err)

// 		bag[color] -= count

// 		if bag[color] < 0 {
// 			// invalid!
// 			return false
// 		}
// 	}

// 	return true
// }

// func isValidRow(row string) int {
// 	// if  valid: return ID
// 	// else, return 0
// 	if row == "" {
// 		// empty is invalid
// 		return 0
// 	}

// 	raw_id, game := split(row, ":")
// 	_, _id := split(raw_id, " ")
// 	id, err := strconv.Atoi(_id)
// 	check(err)

// 	moves := strings.Split(game, ";")

// 	row_valid := true
// 	for _, move := range moves {
// 		move_valid := isValidMove(move)
// 		if !move_valid {
// 			row_valid = false
// 			break
// 		}
// 	}

// 	if row_valid {
// 		return id
// 	} else {
// 		return 0
// 	}
// }

// func run(filename string) int {
// 	dat, err := os.ReadFile(filename)
// 	check(err)

// 	rows := strings.Split(string(dat), "\n")

// 	result := 0

// 	for _, row := range rows {
// 		result += isValidRow(row)
// 	}

// 	return result
// }
// }

func parseRow(row string) int {
	// if  valid: return ID
	// else, return 0
	if row == "" {
		// empty is invalid
		return 0
	}

	_, game := split(row, ":")

	moves := strings.Split(game, ";")

	bag := getEmptyBag()
	for _, move := range moves {
		stones := strings.Split(move, ",")
		for _, stone := range stones {
			splitted := strings.Split(stone, " ")
			_count, color := splitted[1], splitted[2]
			count, err := strconv.Atoi(_count)
			check(err)

			if count > bag[color] {
				bag[color] = count
			}
		}
	}

	result := 1
	for _, count := range bag {
		result *= count
	}

	return result
}

func run2(filename string) int {
	dat, err := os.ReadFile(filename)
	check(err)

	rows := strings.Split(string(dat), "\n")

	result := 0

	for _, row := range rows {
		result += parseRow(row)
	}

	return result
}

File: 2/day2.rs
use std::fs;
use std::collections::HashMap;


type Bag = HashMap<&'static str, u32>;

fn get_bag() -> Bag {
    let mut result = HashMap::new();

    result.insert("red", 12);
    result.insert("green", 13);
    result.insert("blue", 14);
    return result;
}

fn get_empty_bag() -> Bag {
    let mut result = HashMap::new();

    result.insert("red", 0);
    result.insert("green", 0);
    result.insert("blue", 0);
    return result;
}

fn split<'a>(original_string: &'a str, delimiter: &'a str) -> (&'a str, &'a str) {
    let mut parts = original_string.split(delimiter);
    let first_half = parts.next().unwrap_or("");
    let second_half = parts.next().unwrap_or("");

    return (first_half, second_half);
}

fn to_int(txt: &str) -> u32 {
    match txt.parse::<u32>() {
        Ok(n) => n,
        Err(_) => 0,
    }
}

fn valid_round(round: &str, bag: &Bag) -> bool {
    for pair in round.split(",") {
        if pair == "" {
            continue;
        }

        let (_num, color) = split(pair.trim(), " ");
        let num = to_int(_num);

        if bag[color] < num {
            return false;
        }
    }
    return true;
}

fn validate_row(row: &str, bag: &Bag) -> u32 {
    let (_prefix, game) = split(row, ":");

    let (_, _id) = split(_prefix, " ");
    let id = to_int(_id);

    let rounds = game.split(";");

    for round in rounds {
        if !valid_round(round, bag) {
            return 0;
        }
    }

    return id;
}

fn run(filename: &str) -> u32 {
    let contents = fs::read_to_string(filename)
        .expect("Should have been able to read the file");

    let bag = get_bag();

    let rows: Vec<&str> = contents.split("\n").collect();
    let mut result = 0;

    for row in rows {
        result += validate_row(row, &bag);
    }

    return result;
}

fn calculate_bag(row: &str) -> u32 {
    let mut bag = get_empty_bag();

    let (_, game) = split(row, ": ");

    for set in game.split(";") {
        for pair in set.split(",") {
            if pair == "" {
                continue;
            }

            let (_num, color) = split(pair.trim(), " ");
            let num = to_int(_num);

            if num > bag[color] {
                bag.insert(color, num);
            }
        }
    }

    let mut result = 1;
    for (_, count) in bag {
        result *= count;
    }


    return result;
}

fn run2(filename: &str) -> u32 {
    let contents = fs::read_to_string(filename)
        .expect("Should have been able to read the file");

    let rows: Vec<&str> = contents.split("\n").collect();
    let mut result = 0;

    for row in rows {
        result += calculate_bag(row);
    }

    return result;
}

fn main() {
    let result = run("dev.txt");
    println!("result: {r}", r = result);
    assert!(result == 8);

    let pt1_result = run("test.txt");
    println!("part 1: {}", pt1_result);
    assert!(pt1_result == 2679);

    println!("All checks passed for pt1!");

    let result = run2("dev.txt");
    println!("result: {r}", r = result);
    assert!(result == 2286);

    let pt1_result = run2("test.txt");
    println!("part 2: {}", pt1_result);
    assert!(pt1_result == 77607);

    println!("All checks passed for pt2!");
}
